### PR TITLE
Phase 6: OHEM — Online Hard Example Mining for Tandem Configs

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current training uses uniform L1 loss across all 1,322 samples. Tandem configurations are persistently harder (p_tan=29.1 vs p_in=12.1) — the model likely has a subset of samples with consistently high loss that receive the same gradient weight as easy samples. Online Hard Example Mining (OHEM) dynamically tracks a per-sample exponential moving average of loss and upweights persistently hard samples, forcing the model to allocate more capacity to the difficult tandem configurations.

This is distinct from the failed #2000 (OOD-Focused Training) which crashed due to aggressive gradient-based sample selection. OHEM uses a gentle weight multiplier (1.5x) with EMA smoothing — it changes the loss weighting, not the sampling distribution. It also complements the existing `--tandem_ramp` which ramps up tandem sample weights during training; OHEM adds data-adaptive within-tandem reweighting.

**Reference:** OHEM (Shrivastava et al., CVPR 2016). Well-validated in detection (SSD, Faster-RCNN) and dense prediction tasks.

## Instructions

Add OHEM per-sample loss tracking to `cfd_tandemfoil/train.py`. Add three new argparse flags: `--ohem` (bool), `--ohem_weight` (float, default 1.5), `--ohem_threshold_pct` (int, default 75).

**Step 1: Initialize per-sample EMA loss buffer (before training loop)**
```python
if cfg.ohem:
    n_train_samples = len(train_dataset)
    sample_ema_loss = torch.ones(n_train_samples) * 0.5  # on CPU, init to 0.5
    OHEM_DECAY = 0.9
```

**Step 2: Track sample indices in DataLoader**
Ensure the DataLoader returns sample indices alongside data. If the dataset doesn't already return indices, wrap it:
```python
class IndexedDataset(torch.utils.data.Dataset):
    def __init__(self, dataset):
        self.dataset = dataset
    def __getitem__(self, idx):
        return self.dataset[idx], idx
    def __len__(self):
        return len(self.dataset)

if cfg.ohem:
    train_dataset = IndexedDataset(train_dataset)
```

**Step 3: In the training loop, after forward pass, replace loss reduction**
```python
if cfg.ohem:
    with torch.no_grad():
        per_sample_loss = F.l1_loss(pred, target, reduction='none').mean(dim=(1, 2))  # [B]
    # Update EMA on CPU
    sample_ema_loss[batch_indices] = (
        OHEM_DECAY * sample_ema_loss[batch_indices] +
        (1 - OHEM_DECAY) * per_sample_loss.detach().cpu()
    )
    # Compute weights for this batch
    threshold = torch.quantile(sample_ema_loss[batch_indices], cfg.ohem_threshold_pct / 100.0)
    ohem_weights = torch.where(
        sample_ema_loss[batch_indices].to(device) > threshold.to(device),
        torch.full((len(batch_indices),), cfg.ohem_weight, device=device),
        torch.ones(len(batch_indices), device=device)
    )
    loss = (ohem_weights * per_sample_loss).mean()
else:
    loss = F.l1_loss(pred, target)
```

**Step 4: Log OHEM stats each epoch**
```python
if cfg.ohem:
    hard_mask = sample_ema_loss > sample_ema_loss.quantile(cfg.ohem_threshold_pct / 100.0)
    wandb.log({
        "ohem/hard_mean_loss": sample_ema_loss[hard_mask].mean().item(),
        "ohem/easy_mean_loss": sample_ema_loss[~hard_mask].mean().item(),
        "ohem/hard_count": hard_mask.sum().item(),
    }, step=epoch)
```

**Run all 8 GPUs:**
```bash
# 2 baseline seeds (control — run same config without --ohem)
CUDA_VISIBLE_DEVICES=0 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/baseline-s42" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 42 &

CUDA_VISIBLE_DEVICES=1 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/baseline-s73" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 73 &

# OHEM w=1.5, pct=75 (primary config) — 3 seeds
CUDA_VISIBLE_DEVICES=2 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/ohem-w1.5-p75-s42" \
  --ohem --ohem_weight 1.5 --ohem_threshold_pct 75 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 42 &

CUDA_VISIBLE_DEVICES=3 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/ohem-w1.5-p75-s66" \
  --ohem --ohem_weight 1.5 --ohem_threshold_pct 75 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 66 &

CUDA_VISIBLE_DEVICES=4 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/ohem-w1.5-p75-s73" \
  --ohem --ohem_weight 1.5 --ohem_threshold_pct 75 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 73 &

# OHEM w=2.0, pct=75 (stronger weight) — 2 seeds
CUDA_VISIBLE_DEVICES=5 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/ohem-w2.0-p75-s42" \
  --ohem --ohem_weight 2.0 --ohem_threshold_pct 75 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 42 &

CUDA_VISIBLE_DEVICES=6 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/ohem-w2.0-p75-s73" \
  --ohem --ohem_weight 2.0 --ohem_threshold_pct 75 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 73 &

# OHEM w=1.5, pct=90 (top-10% hardest only) — 1 seed
CUDA_VISIBLE_DEVICES=7 python train.py --agent tanjiro \
  --wandb_group phase6/ohem --wandb_name "tanjiro/ohem-w1.5-p90-s42" \
  --ohem --ohem_weight 1.5 --ohem_threshold_pct 90 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed 42 &

wait
```

**Implementation notes:**
- `sample_ema_loss` buffer stays on CPU — only moved to device for weight computation
- Log `ohem/hard_mean_loss` each epoch to watch for instability. If it keeps rising, the weighting is diverging — reduce `ohem_weight`
- Do NOT modify the eval loop — OHEM only affects training
- Run the 2 baselines alongside OHEM to confirm no code regressions

## What to Report

For each run: p_in, p_oodc, p_tan, p_re, val/loss, W&B run ID.

Key comparisons:
1. OHEM w=1.5 vs baseline (same seed) — does p_tan improve?
2. OHEM w=2.0 vs w=1.5 — does stronger weighting help or hurt?
3. OHEM pct=90 vs pct=75 — top-10% vs top-25% upweighting
4. Report `ohem/hard_count` at convergence — what fraction of samples are persistently hard?

## Baseline

Current best (16-seed ensemble, PR #2093):
| Metric | 16-Ensemble | Single-model mean (8 seeds) |
|--------|------------|------------------------------|
| p_in | **12.1** | 13.03 |
| p_oodc | **6.6** | 7.83 |
| p_tan | **29.1** | 30.29 |
| p_re | **5.8** | 6.45 |

**Compare OHEM single-seed results against the single-model mean** (13.03 / 7.83 / 30.29 / 6.45), not the ensemble. A meaningful improvement is >1 std below mean: p_in < 12.56, p_oodc < 7.27, p_tan < 29.35, p_re < 6.29.

W&B project: wandb-applied-ai-team/senpai-v1

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Summary Table — Surface Pressure MAE (lower is better)

| Run | p_in | p_oodc | p_tan | p_re | val_loss | W&B run |
|-----|------|--------|-------|------|----------|---------|
| **baseline-s42** | **13.3** | **7.7** | **30.3** | 6.5 | 0.388 | `tqlbfz9y` |
| **baseline-s73** | **12.9** | 8.1 | 30.5 | 6.6 | 0.392 | `dck4ur8w` |
| ohem-w1.5-p75-s42 | 14.3 | 7.9 | 30.3 | 6.5 | 0.393 | `4x372o82` |
| ohem-w1.5-p75-s66 | 13.4 | 8.0 | 30.7 | 6.6 | 0.392 | `c8v1a5l0` |
| ohem-w1.5-p75-s73 | 13.3 | 8.0 | 30.9 | 6.6 | 0.396 | `6db353lp` |
| ohem-w2.0-p75-s42 | 13.4 | 7.9 | 30.9 | 6.5 | 0.392 | `nzdg07dy` |
| ohem-w2.0-p75-s73 | 13.5 | 8.2 | 31.3 | 6.5 | 0.393 | `0k97472w` |
| ohem-w1.5-p90-s42 | 14.4 | 7.9 | 30.4 | **6.4** | 0.394 | `dng4d3lt` |

**Baseline single-model mean (from PR body):** p_in=13.03, p_oodc=7.83, p_tan=30.29, p_re=6.45

### OHEM Internals (epoch-final)

| Run | hard_count (of 1322) | hard_mean_loss | easy_mean_loss |
|-----|---------------------|----------------|----------------|
| ohem-w1.5-p75 (3 seeds) | 331 (25%) | 0.0140–0.0146 | 0.0052–0.0053 |
| ohem-w2.0-p75 (2 seeds) | 331 (25%) | 0.0140–0.0149 | 0.0053–0.0054 |
| ohem-w1.5-p90 (1 seed)  | 133 (10%) | 0.0191 | 0.0060 |

### Analysis

**OHEM did not improve over baseline.** The hypothesis that upweighting persistently hard samples would improve tandem/hard-case accuracy was not supported.

Key observations:

1. **p_tan (tandem) — no improvement, slight regression.** OHEM w=1.5 p=75 averaged p_tan=30.6 vs baseline 30.4. OHEM w=2.0 was worse at 31.1. The stronger the OHEM weight, the worse p_tan got.

2. **p_in — slight regression.** OHEM w=1.5 p=75 averaged p_in=13.7 vs baseline 13.1. The added sample weighting destabilized in-distribution performance without improving hard cases.

3. **p_re — marginal neutral.** OHEM p=90 had the best single p_re (6.4) but within noise of baseline.

4. **OHEM w=2.0 > w=1.5 is worse.** Stronger upweighting hurt more, suggesting the mechanism is actively harmful, not just neutral.

5. **Interaction with tandem_ramp likely the cause.** The training already has per-sample reweighting via `tandem_ramp` (adaptive_boost) and the hard-node mining (epoch > 30). OHEM adds a *third* layer of reweighting that compounds with these, over-focusing on hard samples and destabilizing the loss landscape. The research agent flagged this risk pre-experiment.

6. **Hard-mean/easy-mean ratio ~2.7x** confirms the EMA is tracking real signal (hard samples are genuinely 2.7x harder). But the model already handles this via existing mechanisms.

### Commands Used

```bash
# Baseline (2 seeds, GPUs 0-1)
CUDA_VISIBLE_DEVICES=0 python train.py --agent tanjiro --wandb_group phase6/ohem --wandb_name "tanjiro/baseline-s42" --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 --seed 42

# OHEM w=1.5 p=75 (3 seeds, GPUs 2-4) — add: --ohem --ohem_weight 1.5 --ohem_threshold_pct 75
# OHEM w=2.0 p=75 (2 seeds, GPUs 5-6) — add: --ohem --ohem_weight 2.0 --ohem_threshold_pct 75
# OHEM w=1.5 p=90 (1 seed, GPU 7) — add: --ohem --ohem_weight 1.5 --ohem_threshold_pct 90
```

Peak memory: ~96 GB (same as baseline — OHEM adds negligible overhead)
All runs: 156-157 epochs, hit 180-min wall-clock timeout.

### Suggested follow-ups

1. **OHEM with tandem_ramp disabled** — test whether OHEM could replace the existing tandem_ramp/adaptive_boost, rather than stacking on top of it. The existing hard-example mechanisms may already be near-optimal.
2. **Curriculum OHEM** — delay OHEM until after warmup (epoch 50+) to avoid interference with the vol_ramp and tandem_ramp curricula. The research agent suggested this and it might avoid the compounding issue.
3. **Per-channel OHEM** — track EMA separately for pressure vs velocity. The current approach mixes channels in surf_per_sample which may dilute the signal.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2101 because Closing — OHEM does not improve any metric. Stronger OHEM weight actively hurts:

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| Baseline (2-seed mean) | 13.1 | 7.9 | 30.4 | 6.6 |
| OHEM w=1.5 p75 (3-seed) | 13.7 (+4.6%) | 8.0 (+1.3%) | 30.6 (+0.7%) | 6.6 (0%) |
| OHEM w=2.0 p75 (2-seed) | 13.5 (+3.1%) | 8.1 (+2.5%) | 31.1 (+2.3%) | 6.5 (-1.5%) |
| OHEM w=1.5 p90 (1-seed) | 14.4 (+9.9%) | 7.9 (0%) | 30.4 (0%) | 6.4 (-3.1%) |

Root cause: OHEM compounds with existing tandem_ramp and adaptive_boost — three layers of reweighting is too much. The OHEM EMA correctly identifies hard samples (2.7x loss ratio) but existing mechanisms already handle this. Stronger OHEM weight → worse metrics, confirming the mechanism is harmful in the presence of existing reweighting. Well-instrumented experiment with clear OHEM logging. Good analysis from the student..
